feat: Ability to Archive Templates

ShadeWyrm opened the pull request:
Repurposes the soft-delete functionality as archive functionality.

Closes #6223 

Deferred issues:
https://github.com/cds-snc/platform-forms-client/issues/6289
https://github.com/cds-snc/platform-forms-client/issues/6290
https://github.com/cds-snc/platform-forms-client/issues/6291
https://github.com/cds-snc/platform-forms-client/issues/6292

timarney commented on lib/integration/prismaConnector.ts:
```diff
@@ -22,7 +22,7 @@ const prismaClientSingleton = () => {
         $allOperations({ args, query }) {
           if ("where" in args) {
-            args.where = { ...args.where, ttl: null };
+            args.where = { ...args.where };
```
@thiessenp-cds 

Looks like you have the most context on the `ttl` here ...

https://github.com/cds-snc/platform-forms-client/pull/5339/files

Concerns with removing?

ShadeWyrm commented on app/(gcforms)/[locale]/(form administration)/forms/components/client/Unarchive.tsx:
```diff
@@ -0,0 +1,52 @@
+import { cloneForm, restoreForm } from "../../actions";
+import { toast, ToastContainer } from "@formBuilder/components/shared/Toast";
+
+export const Unarchive = ({
```
I'd agree, can add as another option in the menu. @samsadasivan - think that makes sense to do?

thiessenp-cds commented on lib/templates.ts:
```diff
@@ -1395,6 +1398,71 @@ export async function deleteTemplate(formID: string): Promise<FormRecord | null>
+        notificationsInterval: true,
+      },
+    })
+    .catch((e) => prismaErrors(e, null));
```
I wonder if we could combine the two prisma queries. So remove the first query and add the isPublished check to the above query. Though I guess the first query is more specific, just finding the template vs. the second that finds that template with no ttl. More specific errors. Just thinking out loud :) Probably good as is.

ShadeWyrm commented on i18n/translations/fr/my-forms.json:
```diff
@@ -3,11 +3,14 @@
   "nav": {
     "drafts": "Ébauches",
     "published": "Publiés",
```
Sorry, not sure I follow - where should that string go in the UI?

anikbrazeau commented on i18n/translations/fr/my-forms.json:
```diff
@@ -3,11 +3,14 @@
     "published": "Publiés",
-    "all": "Tous"
+    "all": "Tous",
+    "archived": "Archivé"
```
```suggestion
    "archived": "Archivés"
```

anikbrazeau commented on i18n/translations/fr/my-forms.json:
```diff
@@ -3,11 +3,14 @@
     "resumeForm": "Reprendre la rédaction de formulaire en cours",
-    "createNewForm": "Créer un nouveau formulaire"
+    "createNewForm": "Créer un nouveau formulaire",
+    "unarchiveForm": "Restaurer le formulaire",
```
```suggestion
    "unarchiveForm": "Désarchiver le formulaire",
```

anikbrazeau commented on i18n/translations/en/my-forms.json:
```diff
@@ -3,11 +3,14 @@
-    "createNewForm": "Create new form"
+    "createNewForm": "Create new form",
+    "unarchiveForm": "Restore form",
+    "duplicatePublishedForm": "Duplicate published form"
```
Question: Why is the action here to duplicate (archived, published) rather than restore?

ShadeWyrm replied:
Initial iteration was that they'd duplicate a published form instead of bringing it back. eg: Fresh and clean copy. Could be changed to whatever implementation we'd like.

anikbrazeau commented on i18n/translations/fr/form-builder.json:
```diff
@@ -477,16 +477,17 @@
+    "okay": "Archiver le formulaire",
     "published": {
       "message1": "La suppression d'un formulaire ne peut être annulée. Votre formulaire publié sera définitivement supprimé de l'Internet, ce qui risque de créer des liens brisés.",
       "message2": "Téléchargez une copie du fichier de formulaire pour vos archives. Vous pourrez ensuite le téléverser à nouveau dans Formulaires GC pour recréer un nouveau formulaire avec de nouveaux liens URL.",
```
```suggestion
      "message1": "L'archivage de ce formulaire publié ne peut être annulée. Votre formulaire publié sera définitivement enlevé de l'Internet, ce qui risque de créer des liens brisés.",
      "message2": "Téléchargez une copie du fichier de formulaire pour vos registres. Vous pourrez ensuite le téléverser à nouveau dans Formulaires GC pour recréer un nouveau formulaire avec de nouveaux liens URL.",
```

anikbrazeau commented on i18n/translations/en/form-builder.json:
```diff
@@ -477,16 +477,17 @@
-      "message2": "Download a copy of the form file for your records. You can then upload it in GC Forms to re-create a new form with new URL links.",
-      "message3": "Before deleting your form, make sure you remove any public links leading to this form. This avoids sending people to a dead end. To stop submissions, you might instead consider closing the form in Settings."
+      "message1": "Archiving will move this published form to the “Archived” section of your forms.",
+      "message2": "Archive and edit your form, to update an existing form instead. The older version will temporarily remain available in the archive folder.",
```
```suggestion
      "message2": "Archive and duplicate your form, to create a new draft from the archived form that can be edited as a new version. The older version will temporarily remain available in the archive folder.",
```

anikbrazeau commented on i18n/translations/fr/my-forms.json:
```diff
@@ -3,11 +3,14 @@
-    "createNewForm": "Créer un nouveau formulaire"
+    "createNewForm": "Créer un nouveau formulaire",
+    "unarchiveForm": "Désarchiver le formulaire",
+    "duplicatePublishedForm": "Dupliquer le formulaire publié"
```
Where does this get duplicated?
 - in what section/folder --> draft or archived? 
 - in what state --> draft or published?

timarney commented:
@ShadeWyrm you mentioned a few times --- believe it was for this PR that there we're some comments / asks that can be broken into another PR --- if so what should be broken out (or was that re: the log PR?)

ShadeWyrm commented:
> @ShadeWyrm you mentioned a few times --- believe it was for this PR that there we're some comments / asks that can be broken into another PR --- if so what should be broken out (or was that re: the log PR?)

That's this one, I'll go through and make them.

ShadeWyrm commented on i18n/translations/fr/my-forms.json:
```diff
@@ -3,11 +3,14 @@
     "resumeForm": "Reprendre la rédaction de formulaire en cours",
-    "createNewForm": "Créer un nouveau formulaire"
+    "createNewForm": "Créer un nouveau formulaire",
+    "unarchiveForm": "Désarchiver le formulaire",
```
Design deferred to another card : https://github.com/cds-snc/platform-forms-client/issues/6289

ShadeWyrm commented on i18n/translations/en/my-forms.json:
```diff
@@ -96,6 +104,9 @@
+    "formDownloadFailed": "Failed to download the form.",
+    "formUnarchiveFailed": "Failed to restore the form."
+  },
+  "archivedNotice": "You can access your archived forms for 30 days.",
```
Deferred to : https://github.com/cds-snc/platform-forms-client/issues/6292

craigzour commented on lib/templates.ts:
```diff
@@ -1395,6 +1398,71 @@ export async function deleteTemplate(formID: string): Promise<FormRecord | null>
+ * @param formID ID of the form template
+ * @returns A boolean status if operation is sucessful
+ */
+export async function restoreTemplate(formID: string): Promise<FormRecord | null> {
```
Now that we have `restoreTemplate` we should probably rename `deleteForm` (with associated logs/errors...) to `archiveForm`.

craigzour commented on lib/templates.ts:
```diff
@@ -1395,6 +1398,71 @@ export async function deleteTemplate(formID: string): Promise<FormRecord | null>
+    throw e;
+  });
+
+  // Check if the form is draft or not.
```
I may be wrong but I feel like this comment does not reflect what is actually being achieved below. Especially considering the thrown error is `TemplateNotFoundError`.

craigzour commented on lib/templates.ts:
```diff
@@ -1395,6 +1398,71 @@ export async function deleteTemplate(formID: string): Promise<FormRecord | null>
+  // Check if the form is draft or not.
+  const template = await prisma.template.findFirstOrThrow({
+    where: {
+      id: formID,
```
If only archived forms can be restored I think we should add a condition on `ttl` not being null to throw an early error.

craigzour commented on lib/integration/prismaConnector.ts:
```diff
@@ -22,7 +22,7 @@ const prismaClientSingleton = () => {
         $allOperations({ args, query }) {
           if ("where" in args) {
-            args.where = { ...args.where, ttl: null };
+            args.where = { ttl: null, ...args.where };
```
With this change, some of the permission checks (related to templates) that are executed on archived forms are throwing errors because Prisma cannot find them since they have a non null `ttl` value.

This is due to the code that retrieves the template in the permission check logic not having `ttl` in the `where` clause:

```typescript
case "FormRecord":
      return prisma.template.findUniqueOrThrow({
        where: {
          id: subject.id,
        },
        select: {
          id: true,
          created_at: true,
          updated_at: true,
          name: true,
          isPublished: true,
          users: {
            select: {
              id: true,
              name: true,
              email: true,
              active: true,
            },
          },
        },
      });
```

I am guessing we can probably add it in there but we will need to make sure it does not break any other part of the code where permission checks on form subject are used.

craigzour commented:
@samsadasivan On the "my forms" main page, should not we also display archived forms when "All" is the selected filter at the top of the page (it is not the case at the moment)?

ShadeWyrm commented:
> @samsadasivan On the "my forms" main page, should not we also display archived forms when "All" is the selected filter at the top of the page (it is not the case at the moment)?

This wouldn't match other apps around the web. (eg: Especially Email)

samsadasivan commented:
Looks great so far! Two small things: 

- A bug with the deleted date: I think it is a month off. It said I deleted the form on "2025-12-12" when I think it meant 11 for the month. Perhaps we would also change the "delete" to "Archived on" :) 

<img width="532" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMUFQREE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--77efbebb135fa4b370d402ac9e2bc759771e3763/image.png" alt="image.png" />

- I archived a published form on October 19th (in the screenshot it says november but I think its october) and its throwing an error when I wanted to duplicate it. I initially thought it was because it was past the 30 days, but I think it may be something else 

<img width="852" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMU1QREE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--293680219c542d2e8321ea46fad55eb7a7d2c4d2/image.png" alt="image.png" />

ShadeWyrm commented:
> Looks great so far! Two small things:
> 
> * A bug with the deleted date: I think it is a month off. It said I deleted the form on "2025-12-12" when I think it meant 11 for the month. Perhaps we would also change the "delete" to "Archived on" :)

Is the date it will be deleted, not the date it was archived. :) 

> 
> * I archived a published form on October 19th (in the screenshot it says november but I think its october) and its throwing an error when I wanted to duplicate it. I initially thought it was because it was past the 30 days, but I think it may be something else


Fixed!~

samsadasivan commented:
My bad on the date haha 

Awesome! Ok ill re-test

craigzour reviewed (approved):
LGTM!
